File: src/main/java/dev/mazurkiewicz/quizer/question/domain/model/Question.java
package dev.mazurkiewicz.quizer.question.domain.model;

import dev.mazurkiewicz.quizer.question.application.AnswerStatus;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;

import java.util.List;

@AllArgsConstructor
@EqualsAndHashCode
@Getter
public class Question {

    private QuestionId id;
    private QuestionContent content;
    private List<Answer> answers;
    private QuestionImage image;
    private QuestionSourceExamDate sourceExamDate;

    public AnswerResult checkAnswer(SelectedAnswer selectedAnswer) {
        AnswerStatus answerStatus = answers.stream()
                .filter(answer -> answer.type() == selectedAnswer.value())
                .findAny()
                .map(Answer::status)
                .orElse(AnswerStatus.INCORRECT);
        return new AnswerResult(answerStatus, correctAnswer().type());
    }

    public Answer correctAnswer() {
        return answers.stream()
                .filter(answer -> answer.status() == AnswerStatus.CORRECT)
                .findFirst()
                .orElseThrow(() -> new IllegalStateException(String.format("Question %d without correct answer!", id.value())));
    }
}

File: src/main/kotlin/io/mazurkiewicz/quizer/question/domain/model/Question.kt
package io.mazurkiewicz.quizer.question.domain.model

import java.util.UUID

data class Question(
    val id: QuestionId,
    val content: QuestionContent,
    val answers: List<Answer>,
    val image: QuestionImage? = null,
    val comment: QuestionComment? = null
) {
    fun checkAnswer(selectedAnswer: SelectedAnswer): AnswerResult {
        val answerFromCafeteria = answers.find { it.type == selectedAnswer.value }
        val selectedAnswerStatus = answerFromCafeteria?.status ?: AnswerStatus.INCORRECT
        return AnswerResult(selectedAnswerStatus, getCorrectAnswer().type)
    }

    private fun getCorrectAnswer(): Answer {
        return answers.find { it.status == AnswerStatus.CORRECT }
            ?: throw IllegalStateException("Question ${id.value} without correct answer!")

    }
}

data class QuestionContent(val value: String)

data class QuestionId(val value: UUID)

data class QuestionComment(val value: String)

data class QuestionImage(val path: String)
